lib/commands/web.js

```diff
@@ -183,11 +195,30 @@ helpers._deleteCookie = async function _deleteCookie (cookieName) {
   await this.executeAtom('execute_script', [script, []]);
 };
 
+helpers.cacheWebElements = function cacheWebElements (rawElements) {
+  if (!(_.isArray(rawElements) || _.isPlainObject(rawElements))) {
+    throw new errors.UnknownError(`Web element identifiers cannot be parsed from ` +
+      JSON.stringify(rawElements));
+  }
+
+  const cacheElement = (el) => {
+    const elId = util.unwrapElement(el);
+    if (!isValidElementIdentifier(elId)) {
+      throw new errors.UnknownError(`The web element identifier ${JSON.stringify(elId)} cannot be parsed`);
+    }
+    const localId = util.uuidV4();
+    this.webElementsCache.set(localId, elId);
```

Comment:
It seems if we can use `elId` by Safari as Appium's return element id _as is_, https://github.com/appium/appium/issues/15230 this issue can be resolved. Or we also can return the same `localId` if the cache already had the same `elId` value.

Hm.. what do you think...?

btw, I observed `:wdc:1628151649325` in current my macOS with Simulator, so potentially newer versions (or Safari browser?) has `wdc` prefix.

Let me check with a real device later.

Reply:
e.g. `:wdc:1628151649325` is by Safari. These two get element request tried to get the same element.

```
[HTTP] --> POST /wd/hub/session/cdf1a2d1-b4b6-49f5-b639-c995eb86c483/element
[HTTP] {"using":"tag name","value":"h1"}
[debug] [W3C (cdf1a2d1)] Calling AppiumDriver.findElement() with args: ["tag name","h1","cdf1a2d1-b4b6-49f5-b639-c995eb86c483"]
[debug] [XCUITest] Executing command 'findElement'
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, name, class name, -ios predicate string, -ios class chain, accessibility id, css selector
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [RemoteDebugger] Executing atom 'find_element' with 'args=["tag name","h1",null]; frames='
[debug] [RemoteDebugger] Executing 'find_element' atom in default context
[debug] [RemoteDebugger] Sending javascript command: '(function(){return function(){var k=this;functi...'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:31058', page '1', target 'page-6' (id: 88): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 88): '"{\"status\":0,\"value\":{\"ELEMENT\":\":wdc:1628151649325\"}}"'
[debug] [RemoteDebugger] Sending to Web Inspector took 147ms
[debug] [RemoteDebugger] Received result for atom 'find_element' execution: {"ELEMENT":":wdc:1628151649325"}
[debug] [W3C (cdf1a2d1)] Responding to client with driver.findElement() result: {"element-6066-11e4-a52e-4f735466cecf":"8d71666c-2dca-42a9-8562-2aa8ca41a641","ELEMENT":"8d71666c-2dca-42a9-8562-2aa8ca41a641"}
[HTTP] <-- POST /wd/hub/session/cdf1a2d1-b4b6-49f5-b639-c995eb86c483/element 200 150 ms - 137
[HTTP]
[HTTP] --> POST /wd/hub/session/cdf1a2d1-b4b6-49f5-b639-c995eb86c483/element
[HTTP] {"using":"tag name","value":"h1"}
[debug] [W3C (cdf1a2d1)] Calling AppiumDriver.findElement() with args: ["tag name","h1","cdf1a2d1-b4b6-49f5-b639-c995eb86c483"]
[debug] [XCUITest] Executing command 'findElement'
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, name, class name, -ios predicate string, -ios class chain, accessibility id, css selector
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [RemoteDebugger] Executing atom 'find_element' with 'args=["tag name","h1",null]; frames='
[debug] [RemoteDebugger] Executing 'find_element' atom in default context
[debug] [RemoteDebugger] Sending javascript command: '(function(){return function(){var k=this;functi...'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:31058', page '1', target 'page-6' (id: 90): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 90): '"{\"status\":0,\"value\":{\"ELEMENT\":\":wdc:1628151649325\"}}"'
[debug] [RemoteDebugger] Sending to Web Inspector took 13ms
[debug] [RemoteDebugger] Received result for atom 'find_element' execution: {"ELEMENT":":wdc:1628151649325"}
[debug] [W3C (cdf1a2d1)] Responding to client with driver.findElement() result: {"element-6066-11e4-a52e-4f735466cecf":"c6c0424e-de22-4b04-b87c-bbd5c1f2f657","ELEMENT":"c6c0424e-de22-4b04-b87c-bbd5c1f2f657"}
[HTTP] <-- POST /wd/hub/session/cdf1a2d1-b4b6-49f5-b639-c995eb86c483/element 200 18 ms - 137
```

Reply:
Changed the cache id generation logic. Could you please check it now @KazuCocoa ?

Reply:
Yey, that can fix appium/appium#15230 in the new element format by Safari.

```
[160] pry(main)> e1 = driver.find_element :tag_name, 'h1'
=> #<Selenium::WebDriver::Element:0x4ae311cb97bb887a id=":wdc:1628180019064">
[161] pry(main)> e = driver.find_element :tag_name, 'h1'
=> #<Selenium::WebDriver::Element:0x4ae311cb97bb887a id=":wdc:1628180019064">
[162] pry(main)> e1 == e
=> true
```

```
[HTTP] --> POST /wd/hub/session/a2a5ac08-fab6-45e0-9994-bf7357afc0fd/element
[HTTP] {"using":"tag name","value":"h1"}
[debug] [W3C (a2a5ac08)] Calling AppiumDriver.findElement() with args: ["tag name","h1","a2a5ac08-fab6-45e0-9994-bf7357afc0fd"]
[debug] [XCUITest] Executing command 'findElement'
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, name, class name, -ios predicate string, -ios class chain, accessibility id, css selector
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [RemoteDebugger] Executing atom 'find_element' with 'args=["tag name","h1",null]; frames='
[debug] [RemoteDebugger] Executing 'find_element' atom in default context
[debug] [RemoteDebugger] Sending javascript command: '(function(){return function(){var k=this;functi...'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:31948', page '1', target 'page-6' (id: 34): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 34): '"{\"status\":0,\"value\":{\"ELEMENT\":\":wdc:1628180019064\"}}"'
[debug] [RemoteDebugger] Sending to Web Inspector took 68ms
[debug] [RemoteDebugger] Received result for atom 'find_element' execution: {"ELEMENT":":wdc:1628180019064"}
[debug] [RemoteDebugger] Web Inspector garbage collected
[debug] [W3C (a2a5ac08)] Responding to client with driver.findElement() result: {"element-6066-11e4-a52e-4f735466cecf":":wdc:1628180019064","ELEMENT":":wdc:1628180019064"}
[HTTP] <-- POST /wd/hub/session/a2a5ac08-fab6-45e0-9994-bf7357afc0fd/element 200 85 ms - 101
[HTTP]
[HTTP] --> POST /wd/hub/session/a2a5ac08-fab6-45e0-9994-bf7357afc0fd/element
[HTTP] {"using":"tag name","value":"h1"}
[debug] [W3C (a2a5ac08)] Calling AppiumDriver.findElement() with args: ["tag name","h1","a2a5ac08-fab6-45e0-9994-bf7357afc0fd"]
[debug] [XCUITest] Executing command 'findElement'
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, name, class name, -ios predicate string, -ios class chain, accessibility id, css selector
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [RemoteDebugger] Executing atom 'find_element' with 'args=["tag name","h1",null]; frames='
[debug] [RemoteDebugger] Executing 'find_element' atom in default context
[debug] [RemoteDebugger] Sending javascript command: '(function(){return function(){var k=this;functi...'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:31948', page '1', target 'page-6' (id: 36): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 36): '"{\"status\":0,\"value\":{\"ELEMENT\":\":wdc:1628180019064\"}}"'
[debug] [RemoteDebugger] Sending to Web Inspector took 15ms
[debug] [RemoteDebugger] Received result for atom 'find_element' execution: {"ELEMENT":":wdc:1628180019064"}
[debug] [RemoteDebugger] Web Inspector garbage collected
[debug] [W3C (a2a5ac08)] Responding to client with driver.findElement() result: {"element-6066-11e4-a52e-4f735466cecf":":wdc:1628180019064","ELEMENT":":wdc:1628180019064"}
[HTTP] <-- POST /wd/hub/session/a2a5ac08-fab6-45e0-9994-bf7357afc0fd/element 200 20 ms - 101
```

Reply:
Tested a few on a real device as well. Seems good.

We should address this change for the next release note... I guess the number of users who face this parsing error will increase if this happens by newer Safari browser versions on host macOS machine.